src/gui/attributetable/qgsfeaturelistview.cpp

```diff
@@ -325,7 +325,11 @@ void QgsFeatureListView::contextMenuEvent( QContextMenuEvent *event )
 
     QgsActionMenu *menu = new QgsActionMenu( mModel->layerCache()->layer(), feature, QStringLiteral( "Feature" ), this );
 
-    emit willShowContextMenu( menu, index );
+    // Index is from feature list model, but we need an index from the
+    // filter model to be passed to listeners, using fid instead would
+    // have been much better in term of bugs (and headaches) but this
+    // belongs to the API unfortunately.
+    emit willShowContextMenu( menu, mModel->mapToSource( index ) );
```

Comment:
This is not nice API :P I think you should remove willShowContextMenu and replace it with your suggestion of using the fid. The existing API is a bug, and should be fixed...

Reply:
I investigated that a bit deeper and there are some actions that require the full selected area (so the full index), I'm afraid the single fid would not be sufficient. 
I'll amend the comment.

Reply:
mmm correcting myself, apparently they get only get item index under the cursor when the context menu is opened, I guess we can retrieve the full selected extent in case we need it anyway.